### ISSUE TITLE
Problems with random number generation

### ISSUE THREAD

**Issue by BlaXpirit** [nim-lang/Nim#1506]
The current implementation of pseudo-random number generation in Nimrod's `math` module has many problems.

First of all, it depends on C's `rand` function, which means different behavior across compilers, library implementations etc. Not that you could guarantee consistent, deterministic behavior even in the same environment, because there is only a single instance of RNG state, letting any code overwrite the seed.

The `rand()` function returns a random integer from 0 to `RAND_MAX`, and this constant also differs. It is 32767 on Windows, meaning that there is only 15 bits of resolution.  
And now for an actual pressing bug: to give a random number from 0 to `max` [the current implementation](https://github.com/Araq/Nimrod/blob/devel/lib/pure/math.nim#L223) basically does `rand() mod max`, and this means you're never going to get a number larger than `RAND_MAX`, even when you're expecting a random number from 0 to, say, a billion. Not to mention [RNG modulo bias](http://google.com/#q=rng+modulo+bias)...

```nimrod
import math

var RAND_MAX {.importc, header: "<stdlib.h>".}: int

for i in 1..1000:
    let n = random(int.high)
    assert n <= RAND_MAX
    echo n

echo int.high # 9223372036854775807 on 64bit
echo RAND_MAX # 32767 on Windows, 2147483647 on Linux
```

The function that returns a random floating point number currently has 48 bits of resolution on Linux and 15 bits on Windows...

The default way to seed (`randomize`) the RNG uses the current time. It is quite a common opinion that doing this is not good. Instead, it is viable to use operating system's cryptographic random byte source.

**Comment by oderwat** [comment]
I think that should be added and math.random + math.randomize should be deprecated. I also think that it should not be in math but in its own module.

**Comment by dom96** [comment]
Shall we adopt @BlaXpirit's random module into the stdlib or keep it as a Nimble package and let it mature?

**Comment by kirbyfan64** [comment]
I would honestly prefer it in stdlib. IMO, a good random number module nowadays is pretty much necessary. Even C++ has one now.

**Comment by BlaXpirit** [comment]
I am not an expert on this topic + my library is too complex. That's a dangerous combination.
I tried my best, but I can't say for sure that there isn't something horribly wrong with it.

Then again, unless something changed in the standard library since I started this issue, trading a 100% chance of something being wrong to a small one seems good. I did include those "chi-square" tests that prove... something.

There is not much more "maturing" that can happen. Personally I couldn't see any changes that would unarguably be beneficial to the library.
I will not return to active development of this library. I will, however, consider every issue raised by others and try to fix it.

The library consists of many files. It would need some work to fit into the standard library.

The bottom line is, if the decision is made to include the library or parts of it into the standard library, I will do the work required from me.
I don't have strong feelings either way.

**Comment by dom96** [comment]
Anybody willing to volunteer their time to move @BlaXpirit's library into the stdlib?

**Comment by Araq** [comment]
The stdlib now uses Xorshift128+, Nim also has MT in its stdlib plus there is @BlaXpirit's nice package. Closing now, everything is fine.
